### PR TITLE
[REVIEW] Add in support for a JAVA API

### PR DESCRIPTION
**Pull request by revans2**
This adds in java APIs to cudf through JNI. To build/test the java APIs make sure you have java 1.8 and maven installed then after building libcudf.so go to the java directory and run `mvn package`

This is joint work with several others including @jlowe @tgravescs @nartal1 @rjafri and @abellina

This does not include CI support yet.  We plan to add that in follow on PRs.  It also is not setup to go to any official maven repository yet.  We plan to do that in follow on work as well.  We are happy to take responsibility for maintaining the java API going forward including reviewing new PRs and bugs.  We will file a separate issue to setup whatever is necessary to make that happen. 

Currently the API is mostly a JNI wrapper around cudf itself.

### CODE REVIEW

**Comment by kkraus14** [comment]
@revans2 the changelog check will fail until you add a brief one-liner to `CHANGELOG.md`

**Comment by revans2** [comment]
> @rapidsai/ops Can we set up an `rapidsai/cudf-java-codeowners` group and tweak the CODEOWNERS file before we merge this? Otherwise this looks good to go to me and further revision can be done in followup PRs smile

@kkraus14 and @rapidsai/ops I just wanted to check if you need anything from me to be able get this setup?

**Comment by harrism** [comment]
> > @rapidsai/ops Can we set up an `rapidsai/cudf-java-codeowners` group and tweak the CODEOWNERS file before we merge this? Otherwise this looks good to go to me and further revision can be done in followup PRs smile
> 
> @kkraus14 and @rapidsai/ops I just wanted to check if you need anything from me to be able get this setup?

I think a list of GitHub usernames you want to be the Java code owners if it's not just you. They will need to be in the rapidsai GitHub org if they are not already.

**Comment by harrism** [comment]
@revans2 this is a large PR, which I can't hope to review in detail. So please let me know if there is anything specific you would like us to review.

File: java/src/main/native/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+    else()
+        message(STATUS "CUDF: Disabling the GLIBCXX11 ABI")
+        set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -D_GLIBCXX_USE_CXX11_ABI=0")
+        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -D_GLIBCXX_USE_CXX11_ABI=0")
+        set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} -Xcompiler -D_GLIBCXX_USE_CXX11_ABI=0")
+    endif(CMAKE_CXX11_ABI)
+endif(CMAKE_COMPILER_IS_GNUCXX)
+
+#set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} -gencode=arch=compute_60,code=sm_60 -gencode=arch=compute_61,code=sm_61")
+set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} -gencode=arch=compute_60,code=sm_60")
```
Comment (harrism) [inline java/src/main/native/CMakeLists.txt]: This file appears to be based on the main cudf/cpp/CMakeLists.txt  Do you have a plan to keep them in sync?

File: java/src/main/native/clang-format.README [hunk 2]
```diff
@@ -0,0 +1,13 @@
+README
```
Comment (harrism) [inline java/src/main/native/clang-format.README]: Are you already using clang-format for the Java side? We haven't added it for C++ yet.

File: java/src/main/native/include/jni_utils.hpp [hunk 3]
```diff
@@ -0,0 +1,1067 @@
...
+class jni_exception : public std::runtime_error {
+public:
+  jni_exception(char const *const message) : std::runtime_error(message) {}
+  jni_exception(std::string const &message) : std::runtime_error(message) {}
+};
+
+/**
+ * @brief throw a java exception and a C++ one for flow control.
+ */
+inline void throw_java_exception(JNIEnv *const env, const char *clazz_name, const char *message) {
```
Comment (harrism) [inline java/src/main/native/include/jni_utils.hpp]: `clazz_name`?

File: java/src/main/native/include/jni_utils.hpp [hunk 4]
```diff
@@ -0,0 +1,1067 @@
...
+    if (wrapped != NULL) {
+      for (int i = 0; i < wrapped->size(); i++) {
+        reset(i, NULL);
+      }
+    }
+  }
+};
+
+/**
+ * @brief RAII for jintArray to be sure it is handled correctly.
```
Comment (harrism) [inline java/src/main/native/include/jni_utils.hpp]: You have multiple RAII wrappers for different types -- similar types like long and int. Could this be replaced by a single template class?

File: java/src/main/native/include/jni_utils.hpp [hunk 5]
```diff
@@ -0,0 +1,1067 @@
...
+    update_caches(index, str);
+  }
+};
+
+/**
+ * Wrapper for gdf_column. This class owns the underlying gdf_column and
+ * release() should be called before assuming ownership of the underlying
+ * gdf_column.
+ */
+class gdf_column_wrapper {
```
Comment (harrism) [inline java/src/main/native/include/jni_utils.hpp]: Will this go away once we have `class cudf::column`?

File: java/src/main/native/clang-format.README (hunk 2 above)
Comment (abellina) [reply java/src/main/native/clang-format.README]: We were using IntelliJ for formatting java, as it is tied in with a fairly popular IDE, but that could change in the future to all use clang-format. Currently clang-format we used only for the cpp code in the java directory, as we tried to get us to be more cudf friendly. When cudf moves to a standard .clang-format file, the idea is for us to move too.

File: java/src/main/native/CMakeLists.txt (hunk 1 above)
Comment (revans2) [reply java/src/main/native/CMakeLists.txt]: If you have some suggestions that would be great.  At this point, we were just going to manually do it, but I know that is far from ideal.

File: java/src/main/native/include/jni_utils.hpp (hunk 3 above)
Comment (revans2) [reply java/src/main/native/include/jni_utils.hpp]: It is a java thing when using reflection.  Happy to change it.

File: java/src/main/native/include/jni_utils.hpp (hunk 4 above)
Comment (revans2) [reply java/src/main/native/include/jni_utils.hpp]: The JNI API is not really a C++ API.  It is a C API with some magic that lets you call some functions like C++ methods.  So we would need a way to drop in the proper function for each subclass `GetLongArrayElements`, `GetIntArrayElements`, etc in each instantiation of the template.  I think we could do it, but it is not simple.  Happy to explore it more in a follow-on issue.

File: java/src/main/native/include/jni_utils.hpp (hunk 5 above)
Comment (revans2) [reply java/src/main/native/include/jni_utils.hpp]: Yes.  I sure hope so.

**Comment by revans2** [comment]
> @revans2 this is a large PR, which I can't hope to review in detail. So please let me know if there is anything specific you would like us to review.

@harrism I would love feedback on our C++ code.  Our team hasn't done serious C++ development for several years and that was for older C++ releases, so if you see things that we can do better it would be great to point them out and we can look at fixing them now or in follow on issues.

**Comment by revans2** [comment]
@kkraus14 and @rapidsai/ops the people to add to the group are the ones I listed as contributors to the patch.

@jlowe @tgravescs @nartal1 @rjafri @abellina and myself @revans2

File: java/src/main/native/include/jni_utils.hpp (hunk 4 above)
Comment (revans2) [reply java/src/main/native/include/jni_utils.hpp]: I made it work.  Not sure it is the cleanest implementation.

f9a73fb and 45eeeed

File: java/src/main/native/CMakeLists.txt (hunk 1 above)
Comment (harrism) [reply java/src/main/native/CMakeLists.txt]: There are basically two approaches: `add_subdirectory` and `include`. So we could `add_subdirectory` from the top-level `CMakeLists.txt` like we do for the `cpp/tests/CMakeLists.txt`. Or we could factor common functionality out of the top-level one and `include` it from this one and a new one that we put at the top level.

In either case, anything in an included or add_subdirectory Lists will get run, so you have to be careful to not duplicate targets.

Comment (harrism) [reply java/src/main/native/CMakeLists.txt]: This can be a separate PR. Until it happens though you'll have to keep an eye on cmake changes.

**Comment by harrism** [comment]
> @harrism I would love feedback on our C++ code.

You managed to abstract some of your array accessors to reduce code duplication, which helps. Other than that nothing jumped out at me. I wondered if some of the switch statements could use `type_dispatcher` but it wasn't obvious to me that they could. @jrhemstad is likely to have comments.

**Comment by revans2** [comment]
@kkraus14 @jrhemstad is there anything else I need to do for this?  There are some other changes I would like to get into the java API before the 0.8 release, but I didn't want to modify this PR significantly once it is up.

**Comment by kkraus14** [comment]
I defer to @rapidsai/ops here. I have no qualms from my perspective.

File: java/src/main/native/CMakeLists.txt (hunk 1 above)
Comment (saudet) [reply java/src/main/native/CMakeLists.txt]: FYI, that's what JavaCPP is for: It's a tool that generates all that JNI boilerplate code!
See issue #563

**Comment by archenroot** [comment]
Guys, I wonder why you didn't choose JavaCPP Presets to do this, its quite comphrensive framework to provide more "Javoized" JNI interfaces:

https://github.com/bytedeco/javacpp-presets

I will work on these interfaces, just waited for some time until cudf gets more stable (so don't need every 2 weeks to change api)....

File: java/src/main/native/CMakeLists.txt (hunk 1 above)
Comment (archenroot) [reply java/src/main/native/CMakeLists.txt]: Yeah, I also suggest to turn the way to JavaCPP utilization, will help a lot.

**Comment by revans2** [comment]
@archenroot We explored several different options for generating a Java API and it really came down to a few things.

First was performance.  JNI is horribly slow.  We wanted to reduce its use as much as possible, make sure that we were only using it on the control path and not the data path, and when it was used we didn't go back and forth between Java and native all the time. This is why we are using Unsafe to read and write data into the raw buffers.  At some point, we may look into using these tools to automate some things, like generating enums, that are a pain to keep in sync and don't have any real impact on performance.

The second was memory management.  Most tools that auto-generate java classes rely on the garbage collector to release the native memory that they point to.  This is fine if you are doing simple things with little data, or if the process is short lived.  If we want to point to a 5GB tables sitting in GPU memory that is going to go through multiple transformations creating large intermediate results I don't want to hope that the Java GC happens to run at the right time.  We wanted clear lifetimes and complete control, which is why we are using `AutoCloseable` coupled with reference counting and then backing it up with Java GC.

The final one was the API itself.  https://github.com/bytedeco/javacpp-presets/blob/master/cuda/samples/MNISTCUDNN.java does not look like a Java API.  It looks like a C API that someone is calling from java, i.e. `FloatPointer`. If it weren't for the other two issues we could just write a cleaner wrapper around the generated API.  But all of it together led us to do what we did.

**Comment by saudet** [comment]
@revans2 JavaCPP does use `Unsafe` and `AutoCloseable` for exactly the same reasons you mention:

http://bytedeco.org/news/2014/12/23/third-release/
http://bytedeco.org/news/2015/10/25/moving-to-java-se-7/
http://bytedeco.org/news/2018/07/17/bytedeco-as-distribution/

Of course, the generated Java interface is going to look like the original C/C++ API, but the point is to build on that, instead of JNI.

**Comment by revans2** [comment]
@archenroot Will take a closer look at it.  Thanks for the suggestion.

**Comment by archenroot** [comment]
Hi, any decision taken on the path?

**Comment by razajafri** [comment]
@archenroot as per our conversation in the [https://github.com/bytedeco/javacpp-presets/issues/665](url) 
The pain point right now seems to be the way to organize headers that need to be in a specific order. I also would like to get a better understanding of whether we need to add Arrow presets as a pre-requisite to creating the cudf presets. I am fine with continuing the conversation here although I think it will be beneficial to others if we continue in javacpp

**Pull request merged**